atecc608a_se.c

```diff
@@ -0,0 +1,227 @@
+    case ATCA_DATA_ZONE_LOCKED:
+    case ATCA_NOT_LOCKED:
+    case ATCA_WAKE_FAILED:
+    case ATCA_STATUS_UNKNOWN:
+    case ATCA_STATUS_ECC:
+    case ATCA_STATUS_SELFTEST_ERROR:
+    case ATCA_CHECKMAC_VERIFY_FAILED:
+    case ATCA_PARSE_ERROR:
+    case ATCA_FUNC_FAIL:
+    case ATCA_GEN_FAIL:
+    case ATCA_EXECUTION_ERROR:
+    case ATCA_HEALTH_TEST_ERROR:
+    case ATCA_INVALID_SIZE:
+    default:
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+}
+
+static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
+                                                uint8_t *p_data, size_t data_size,
+                                                size_t *p_data_length)
+{
+    const size_t key_data_len = 65;
+    const uint16_t slot = key;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    if (data_size < key_data_len)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    ATCAB_INIT();
+
+    /* atcab_get_pubkey returns concatenated x and y values, and the desired 
+       format is 0x04 + x + y. We start at &p_data[1] and add a 0x04 at p_data[0]. */
+    ASSERT_SUCCESS(atcab_get_pubkey(slot, &p_data[1]));
+
+    p_data[0] = 4;
+    *p_data_length = key_data_len;
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                              psa_algorithm_t alg,
+                                              const uint8_t *p_hash,
+                                              size_t hash_length,
+                                              uint8_t *p_signature,
+                                              size_t signature_size,
+                                              size_t *p_signature_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* We can only do ECDSA on SHA-256 */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
+    {
+        /* The driver only supports signing things of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (signature_size < ATCA_SIG_SIZE)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    ATCAB_INIT();
+
+    /* Signature will be returned here. Format is R and S integers in
+     * big-endian format. 64 bytes for P256 curve. */
+    ASSERT_SUCCESS(atcab_sign(key_id, p_hash, p_signature));
+         
+    *p_signature_length = ATCA_SIG_SIZE;
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_asymmetric_sign - signature size %d:\n", *p_signature_length);
+    atcab_printbin_sp(p_signature, *p_signature_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+
+#define PSA_ATECC608A_LIFETIME 0xdeadbeefU
```

Comment:
This should probably be in the header file, so that applications can ask for a slot from the SE using it.

Reply:
This can be changed later on, when full (or just better) support of the SE is added.